web/packages/teleport/src/Users/UserAddEdit/TraitsEditor.tsx

```diff
@@ -0,0 +1,271 @@
+/**
+ * Teleport
+ * Copyright (C) 2024  Gravitational, Inc.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+import React, { useEffect, Dispatch, SetStateAction } from 'react';
+import { ButtonBorder, Box, Flex, Text, ButtonIcon } from 'design';
+import { Add, Trash } from 'design/Icon';
+import { FieldSelectCreatable } from 'shared/components/FieldSelect';
+import { Option } from 'shared/components/Select';
+import { requiredField, requiredAll } from 'shared/components/Validation/rules';
+import { Attempt } from 'shared/hooks/useAttemptNext';
+
+import { AllUserTraits } from 'teleport/services/user';
+
+/**
+ * traitsPreset is a list of system defined traits in Teleport.
+ * The list is used to populate traits key option.
+ */
+const traitsPreset = [
+  'aws_role_arns',
+  'azure_identities',
+  'db_names',
+  'db_roles',
+  'db_users',
+  'gcp_service_accounts',
+  'host_user_gid',
+  'host_user_uid',
+  'kubernetes_groups',
+  'kubernetes_users',
+  'logins',
+  'windows_logins',
+];
+
+/**
+ * TraitsEditor supports add, edit or remove traits functionality.
+ * @param allTraits all traits pre-configured for user.
+ * @param attempt attempt is Attempt status.
+ * @param configuredTraits holds traits configured for user in current editor.
+ * @param setConfiguredTraits sets user traits in current editor.
+ */
+export function TraitsEditor({
+  allTraits,
+  attempt,
+  configuredTraits,
+  setConfiguredTraits,
+}: TraitEditorProps) {
+  useEffect(() => {
+    const newTraits = traitsToTraitsOption(allTraits);
+    setConfiguredTraits(newTraits);
+  }, [allTraits]);
+
+  function handleInputChange(i: InputOption) {
+    let newTraits = [...configuredTraits];
+    if (i.labelField === 'traitValues') {
+      let traitValue: Option[] = i.option as Option[];
```

Comment:
This is needed because `InputOption.option` accepts both `Option | Option[];`. The alternative would be to define seperate type for input options
```
type InputOption = {
  labelField: 'traitKey';
  option: Option;
  index: number;
};

type InputOptionArray = {
  labelField: 'traitValues';
  option: Option[];
  index: number;
};
```

and do `function handleInputChange(i: InputOption | InputOptionArray) {`. But that still needs to specify type on handlechange 
```
onChange={e => {
    handleInputChange({
    option: e as Option[],
```
Not sure which one it better typed though! Any suggestions?

Reply:
i would do exactly as you proposed, i think we are lacking typing on `FieldSelectCreatable` to specify like a generic type for options